### PR TITLE
Feature/ha discovery

### PR DESCRIPTION
**Pull request by dominikandreas**
This is a work in progress, but wanted to get some early feedback. The main contributions of this PR are:

- Implemented support for home assistant discovery (https://github.com/DennisOSRM/hms-mqtt-publisher/issues/17) 
- Add extra sensors (https://github.com/DennisOSRM/hms-mqtt-publisher/issues/14) 

Other notable changes / additions:
- added vscode devcontainer support and debug launch configuration to simplify development for future contributors
- add serde dependency for json serialization
- version bump to 0.2.0 because of breaking changes (mqtt paths change)
- refactor log init into its own module
- add mock_state function to inverter for testing purposes

Note that this is my first time developing in rust and there's probably a lot of things to improve / things that an experienced rust developer would've probably done differently. I'm happy to receive some constructive feedback and improve where necessary :-)

TODO:
- update home assistant addon once PR is merged / docker build is ready
- make home assistant prefix configurable (https://github.com/DennisOSRM/hms-mqtt-publisher/issues/25)

### CODE REVIEW

**Comment by DennisOSRM** [comment]
Thanks for the PR. It will take a couple of days before I can get around to reviewing this. 

From a quick glance, it seems this PR contains a couple of things that likely should be seperate PRs, and the testing code perhaps should not bleed into the implementation. That being said, this isn't a comprehensive review and just an initial impression. 

Thanks for the contribution, please keep up the great work.

File: .devcontainer/Dockerfile [hunk 1]
```diff
@@ -0,0 +1 @@
+FROM rust:buster
```
Comment (DennisOSRM) [inline .devcontainer/Dockerfile:1]: The files in .devcontainer should not be part of this PR.

File: .vscode/launch.json [hunk 2]
```diff
@@ -0,0 +1,21 @@
+{
```
Comment (DennisOSRM) [inline .vscode/launch.json:1]: The files in `.vscode/*` should not be part of this PR.

File: src/inverter.rs [hunk 3]
```diff
@@ -39,7 +40,42 @@ impl<'a> Inverter<'a> {
         }
     }
 
-    pub fn update_state(&mut self) -> Option<HMSStateResponse> {
+    pub fn mock_state(&mut self) -> Option<HMSStateResponse> {
```
Comment (DennisOSRM) [inline src/inverter.rs:43]: The test code should be separated from the implementation. This could be done in a tool simulating the behavior of an actual inverter. Issue #11 is tracking such a feature. This should be dropped from the PR.

File: src/main.rs [hunk 4]
```diff
@@ -62,11 +52,15 @@ fn main() {
...
 
     loop {
-        if let Some(r) = inverter.update_state() {
+        if let Some(r) = inverter.update_state(cli.test) {
             mqtt.publish(&r);
         }
 
         // TODO: this has to move into the Inverter struct in an async implementation
-        thread::sleep(Duration::from_millis(REQUEST_DELAY));
+        if !cli.test{
```
Comment (DennisOSRM) [inline src/main.rs:60]: See above for the comment that API simulation should implemented in #11.

File: src/mqtt.rs [hunk 5]
```diff
@@ -13,14 +15,221 @@ pub struct Mqtt {
...
 }
 
+
+/// `HMSStateResponse` is a struct that contains the data from the inverter.
+///
+/// Provide utility functions to extract data from the struct.
+impl HMSStateResponse {
+    fn get_model(&self) -> String {
+         // TODO: identify model from dtu_sn
+        "HMS-800W-T2".to_string()
```
Comment (DennisOSRM) [inline src/mqtt.rs:25]: This is not necessarily the inverter. While tested, the tool likely works with all devices of the HMS-XXXX-2T series.

File: src/mqtt.rs [hunk 6]
```diff
@@ -13,14 +15,221 @@ pub struct Mqtt {
...
+        };
+    }
+
+    fn get_inverter(&self, pv_port: i32) -> InverterState {
+        self.inverter_state.iter().find(
+            |inv| inv.port_id == pv_port
+        ).unwrap_or(&InverterState::default()).clone()
+    }
+
+    fn to_json_payload(&self) -> serde_json::Value {
```
Comment (DennisOSRM) [inline src/mqtt.rs:53]: consider using the [protobuf-json](https://crates.io/crates/protobuf-json) crate, which makes this function a one-liner basically.

File: src/mqtt.rs [hunk 7]
```diff
@@ -13,14 +15,221 @@ pub struct Mqtt {
...
+        }
+
+        json
+    }
+}
+
+/// `DeviceConfig` is used to define the configuration for a Home Assistant device
+/// in the MQTT discovery protocol and is used to group entities together.
+/// 
+#[derive(Serialize, Clone)]
```
Comment (DennisOSRM) [inline src/mqtt.rs:89]: consider moving the struct into a separate source file

File: src/mqtt.rs [hunk 8]
```diff
@@ -13,14 +15,221 @@ pub struct Mqtt {
...
+pub struct DeviceConfig {
+    name: String,
+    model: String,
+    identifiers: Vec<String>,
+    manufacturer: String,
+    sw_version: String,  // Software version of the application that supplies the discovered MQTT item.
+}
+
+
+impl DeviceConfig {
```
Comment (DennisOSRM) [inline src/mqtt.rs:99]: consider moving the struct into a separate source file

File: src/mqtt.rs [hunk 9]
```diff
@@ -13,14 +15,221 @@ pub struct Mqtt {
...
+    pub fn new(name: String, model: String, identifiers: Vec<String>) -> Self {
+        Self {
+            name,
+            model,
+            identifiers,
+            manufacturer: "Hoymiles".to_string(),
+            /// Rust compiler sets the CARGO_PKG_VERSION environment from the Cargo.toml . 
+            sw_version: env!("CARGO_PKG_VERSION").to_string(),
```
Comment (DennisOSRM) [inline src/mqtt.rs:107]: This is arguably not part the version of the inverter device

File: src/mqtt.rs [hunk 10]
```diff
@@ -13,14 +15,221 @@ pub struct Mqtt {
...
+/// in the MQTT discovery protocol.
+/// 
+/// More information about the MQTT discovery protocol can be found here:
+/// https://www.home-assistant.io/docs/mqtt/discovery/
+/// 
+/// More information about the Home assistant sensor entities can be found here:
+/// https://developers.home-assistant.io/docs/core/entity/sensor/
+///
+#[derive(Serialize)]
+pub struct SensorConfig {
```
Comment (DennisOSRM) [inline src/mqtt.rs:122]: consider moving this to a separate source file

File: src/mqtt.rs [hunk 11]
```diff
@@ -13,14 +15,221 @@ pub struct Mqtt {
...
+    }
+
+    pub fn efficiency(state_topic: &str, device_config: &DeviceConfig,  name: &str, key: &str) -> Self {
+        Self::new_sensor(state_topic, &device_config, key, name, "", "%")
+    }
+
+}
+
+
+fn create_sensor_configs(hms_state: &HMSStateResponse, state_topic: &str) -> Vec<SensorConfig> {
```
Comment (DennisOSRM) [inline src/mqtt.rs:188]: should this be a `::from(...)` member function rather than a free factory?

File: src/mqtt.rs [hunk 12]
```diff
@@ -42,35 +251,41 @@ impl Mqtt {
...
     fn publish(&mut self, hms_state: &HMSStateResponse) {
-        debug!("{hms_state}");
-
-        let pv_current_power = hms_state.pv_current_power as f32 / 10.;
-        let pv_daily_yield = hms_state.pv_daily_yield;
-
-        self.client
-            .subscribe("hms800wt2/pv_current_power", QoS::AtMostOnce)
-            .unwrap();
-        match self.client.publish(
```
Comment (DennisOSRM) [inline src/mqtt.rs:57]: this seems like a breaking change.

**Review by DennisOSRM — CHANGES_REQUESTED** [review]
Thanks again for filing this PR. I think it contains very interesting and worthwhile ideas. Since the PR is rather big and contains several unrelated changes, I'd suggest to do the following:


- [ ] remove the development related files and directories
- [ ] the testing code should be done in a distinct PR without having to branch between test and non-test code. I.e. the implementation should be oblivious to the fact whether it is run in a test environment or not.
- [ ] move structs into their own source file
- [ ] consider using `protobuf_json` crate instead of the rolling your own conversion to json
- [ ] implement the auto discovery in a non-breaking way

While the above may seem like a lot, I think the PR is a great first stab into the right direction. The intent for the PR is great, please keep up the good work.

**Comment by dominikandreas** [comment]
Thanks for your comprehensive review!

> remove the development related files and directories

I will create a separate PR for this

> the testing code should be done in a distinct PR without having to branch between test and non-test code. 

Good point. Didn't have time to address this properly so just hacked in something quickly that allowed me to do some minimal testing during the addon development. I will remove the testing related code for now

> move structs into their own source file

Will do. Wasn't sure how to structure it or whether it's necessary to split the mqtt implementation up in two parts, one for home assistant, one without home assistant support.

> consider using protobuf_json crate instead of the rolling your own conversion to json

Thought about this as well, but since the values are scaled during the serialization, I decided to just do it the manual way. I assume the protobuf schemas won't often if at all, so I hope that's okay. 

>  implement the auto discovery in a non-breaking way

The breaking change is unfortunate. But apart from home assistant support, another upside is that it would allow utilizing multiple inverters in the future (part of the serial number is added to the topic). Another part of the breaking change is that the whole update is sent json-encoded to a single topic instead of splitting it to multiple topics. This should make it more efficient in combination with home assistant (and possibly other endpoints)

**Comment by DennisOSRM** [comment]
> The breaking change is unfortunate. But apart from home assistant support, another upside is that it would allow utilizing multiple inverters in the future (part of the serial number is added to the topic). Another part of the breaking change is that the whole update is sent json-encoded to a single topic instead of splitting it to multiple topics. This should make it more efficient in combination with home assistant (and possibly other endpoints)

That's a good point. Thanks for the clarification. I think the json-encoded single topic could be done independently, is that right? If so, we could first add things, and consider the breaking change at a later point. Would this make sense?

**Comment by dominikandreas** [comment]
Yes, we could change the home assistant device configurations and publish on separate topics to delay the breaking change. But if we're certain it's going to happen sooner or later (and I think supporting multiple inverters is a enough good reason), delaying the breaking change will only cause more users to be affected. 

While we're at it we should also remove the "800W-2T" references, e.g. from the topic, to prevent a future breaking change.

**Comment by DennisOSRM** [comment]
Ok, that's a convincing argument. Let's do the breaking change now, and aim at the same time to minimize the size of this PR.

File: src/mqtt.rs (hunk 5 above)
Comment (dominikandreas) [reply src/mqtt.rs]: Right, since we currently don't have a way to determine the exact model name, I've removed the mention of 800W-2T

File: src/mqtt.rs (hunk 9 above)
Comment (dominikandreas) [reply src/mqtt.rs:107]: No, but I thought it makes sense to log the version of the publisher here (we can also change it in the future in case we figure out a way to extract the sw version of the inverter)

**Comment by dominikandreas** [comment]
I think I addressed all of the review comments and fixed some additional bugs. Here's an example of the provided data in home assistant:

image

In order to enable these changes in the addon, we first need an image on docker hub that I can reference in the Dockerfile of the home assistant addon. I have a build running right now, but it's taking a while. I will hopefully push another update tomorrow when it's done

**Comment by DennisOSRM** [comment]
Very cool. Please have a bit of patience, since it will again take a couple of days before I can get to it.

File: Cargo.toml [hunk 13]
```diff
@@ -1,6 +1,6 @@
 [package]
 name = "hms-mqtt-publish"
-version = "0.1.0"
+version = "0.2.0"
```
Comment (DennisOSRM) [inline Cargo.toml:3]: minor nitpick. let's leave this at 0.1.0 for now. The version will be increased as part of the release process.

File: .gitignore [hunk 14]
```diff
@@ -14,4 +14,4 @@ Cargo.lock
 *.pdb
 
 # Ignore VS Code user settings
-.vscode/
\ No newline at end of file
+.vscode/
```
Comment (DennisOSRM) [inline .gitignore:17]: please drop this white-space change

File: src/inverter.rs [hunk 15]
```diff
@@ -38,7 +38,7 @@ impl<'a> Inverter<'a> {
             info!("Inverter is {new_state:?}");
         }
     }
-
+
```
Comment (DennisOSRM) [inline src/inverter.rs:41]: please run `cargo fmt` to ensure proper formatting and removal of stray whitespace

File: src/mqtt.rs [hunk 16]
```diff
@@ -42,35 +47,140 @@ impl Mqtt {
 
         Self { client }
     }
+
+    fn publish_json(&mut self, topic: &str, payload: serde_json::Value) {
+        debug!("Publishing to {topic} with payload {payload}");
+
+        let payload = serde_json::to_string(&payload).unwrap();
```
Comment (DennisOSRM) [inline src/mqtt.rs:54]: why not use the `protobuf_json` crate to publish the original response?

File: src/mqtt.rs [hunk 17]
```diff
@@ -42,35 +47,140 @@ impl Mqtt {
...
+            |port| port.pv_power as f32
+        ).sum();
+        if total_module_power > 0.0 {
+            self.pv_current_power as f32 / total_module_power * 100.0
+        } else { 
+            0.0
+        }
+    }
+
+    fn to_json_payload(&self) -> serde_json::Value {
```
Comment (DennisOSRM) [inline src/mqtt.rs:118]: one thing that I haven't fully understood, yet of this is a json schema that HA requires. Is there any formal requirement, and where can I find some information on this?

File: src/mqtt_schemas.rs [hunk 18]
```diff
@@ -0,0 +1,114 @@
+
```
Comment (DennisOSRM) [inline src/mqtt_schemas.rs:1]: the file should be named after the struct in it, or the most important one.

File: src/protos/RealData.proto [hunk 19]
```diff
@@ -12,7 +12,7 @@ message InverterState {
   int64 inv_id = 1;
   int32 port_id = 2;
   int32 grid_voltage = 3;     // [V], factor 0.1
-  int32 grid_freq = 4;        // [Hz], factor 0.1
+  int32 grid_freq = 4;        // [Hz], factor 0.01
```
Comment (DennisOSRM) [inline src/protos/RealData.proto:15]: arguably, the protobuf change should go in independently of this PR.

File: src/main.rs [hunk 20]
```diff
@@ -31,19 +31,7 @@ struct Cli {
 static REQUEST_DELAY: u64 = 30_500;
 
 fn main() {
-    Builder::new()
```
Comment (DennisOSRM) [inline src/main.rs:34]: can we move the non-functional change into it's own PR?

**Review by DennisOSRM — COMMENTED** [review]
- [ ] run `cargo fmt`
- [ ] move non-functional changes into distinct PRs to separate concerns

This is looking really close. Couple of remarks on the PR that shouldn't take too long to resolve. Then we can get this in.

To give some perspective of where I am taking the architecture of the tool are probably helpful. Similar to the idea of supporting several inputs, I'd like the tool to support several output channels (for a lack of a better name). I.e. there's a simple MQTT publisher, there's the HA autodiscovery and publishing of this PR, and other outputs like InfluxDB and perhaps others. The user will be able to configure where to publish.

Once this PR is ready, I want to start implementing these channels by building right on top of this branch and then merge the two in short succession once both are ready.

**Comment by DennisOSRM** [comment]
I did a very hasty implementation of output channels (cf. https://github.com/DennisOSRM/hms-mqtt-publisher/pull/27#pullrequestreview-1746689155) on top of your PR: https://github.com/DennisOSRM/hms-mqtt-publisher/pull/34/commits/7460c0e6194d36540d5682457d884d2e632e5074

File: src/mqtt.rs [hunk 21]
```diff
@@ -1,9 +1,12 @@
 use std::{thread, time::Duration};
```
Comment (DennisOSRM) [inline src/mqtt.rs:1]: in light of the _output channel_ feature (see below) it probably makes sense to rename this file to `home_assistent.rs` since it will implement the output channel to a HA instance.

**Review by DennisOSRM — APPROVED** [review]
Will merge changes and do a couple of minor changes on top. Thanks so much for the contribution

File: src/mqtt.rs (hunk 17 above)
Comment (dominikandreas) [reply src/mqtt.rs:118]: Home assistant requries two parts: config messages and state messages. The config messages define the sensors and corresponding state topics to listen for state messages. 

The state messages don't require any specific schema. Rather the configuration messages contain a "value_template" (`SensorConfig.value_template`), which is used to extract the state value from a json-encoded message sent to the "state_topic" (`SensorConfig.state_topic`). The template language used is jinja.

We've defined the template for each sensor as `let value_template = format!("{{{{ value_json.{} }}}}", unique_id);`, whereas `value_json` would correspond to the json-serialized `HMSStateResponse` and `unique_id` is defined to match a corresponding attribute in `HMSStateResponse`

File: src/mqtt.rs (hunk 16 above)
Comment (dominikandreas) [reply src/mqtt.rs:54]: did you mean regarding line 77? The `hms_state.to_json_payload` includes the scaling and the additional efficiency attribute. wasn't sure how to do that without manually implementing the json conversion

File: src/mqtt_schemas.rs (hunk 18 above)
Comment (dominikandreas) [reply src/mqtt_schemas.rs:1]: yes, wasn't happy with the naming either, but couldn't think of anything else. Feel free to suggest an alternative or directly rename it to something else

File: src/mqtt.rs (hunk 16 above)
Comment (DennisOSRM) [reply src/mqtt.rs:54]: Makes sense.

**Comment by dominikandreas** [comment]
Sorry I was preoccupied with my day job and family. I'll have a look at the output channels PR

**Comment by DennisOSRM** [comment]
No worries. Life comes first, hacking on code second. 😉
